swift_kernel.py

```diff
@@ -823,6 +837,184 @@ def is_valid_dependency(path):
+            swift_module = os.path.join(bazel_bin_dir, swift_module)
+            if os.path.exists(swift_module):
+                swift_module_search_paths.add(os.path.dirname(swift_module))
+                self.send_response(self.iopub_socket, 'stream', {
+                    'name': 'stdout',
+                    'text': 'swiftmodule: %s\n' % swift_module
+                })
+        result = subprocess.check_output(["bazel", "query","kind(cc_library, deps(//.ibzlnb/%s:libjupyterInstalledPackages.so))" % tmpdir], env=env).decode("utf-8")
+        # Process *.modulemaps
+        for dep in result.split("\n"):
+            dep = dep.strip()
+            if len(dep) == 0:
+                continue
+            if dep[0] == '@':
+                parts = dep[1:].split("//")
+                modulemaps = '/'.join(filter(lambda x: len(x) > 0, ['external', parts[0]] + parts[1].split(":"))) + ".modulemaps"
+            else:
+                parts = dep.split("//")
+                modulemaps = '/'.join(filter(lambda x: len(x) > 0, [parts[0]] + parts[1].split(":"))) + ".modulemaps"
+            modulemaps = os.path.join(bazel_bin_dir, modulemaps)
+            if os.path.exists(modulemaps):
+                swift_module_search_paths.add(os.path.dirname(modulemaps))
+                self.send_response(self.iopub_socket, 'stream', {
+                    'name': 'stdout',
+                    'text': 'modulemaps: %s\n' % modulemaps
+                })
+        lib_filename = os.path.join(bazel_bin_dir, '.ibzlnb', tmpdir, 'libjupyterInstalledPackages.so')
+        self.swift_module_search_paths = list(swift_module_search_paths)
+
+        # == dlopen the shared lib ==
+
+        self.send_response(self.iopub_socket, 'stream', {
+            'name': 'stdout',
+            'text': 'Initializing Swift...\n'
+        })
+        self._init_swift()
+
+        dynamic_load_code = textwrap.dedent("""\
+            import func Glibc.dlopen
+            import var Glibc.RTLD_NOW
+            dlopen(%s, RTLD_NOW)
+        """ % json.dumps(lib_filename))
+        dynamic_load_result = self._execute(dynamic_load_code)
+        if not isinstance(dynamic_load_result, SuccessWithValue):
+            raise PackageInstallException(
+                    'Install Error: dlopen error: %s' % \
+                            str(dynamic_load_result))
+        if dynamic_load_result.value_description().endswith('nil'):
+            raise PackageInstallException('Install Error: dlopen error. Run '
+                                        '`String(cString: dlerror())` to see '
+                                        'the error message.')
+
+        self.send_response(self.iopub_socket, 'stream', {
+            'name': 'stdout',
+            'text': 'Installation complete!\n'
+        })
+        # Switch back to whatever we previous from.
+        os.chdir(pwd)
+        self.already_installed_packages = True
+
+
```

Comment:
(Optional) This `SwiftKernel` megaclass is becoming pretty unweildy and it's getting hard to keep track of what methods are used for what functionality. Splitting the installation functionality into a different class in a different file could be a good place to start fixing this :)